### PR TITLE
Add multiple entry points support

### LINKED ISSUE #113: Listen on both HTTP and HTTPS
It'd be pretty rad if we could listen on both HTTP and HTTPS at the same time.

### PR DESCRIPTION
**Pull request by emilevauge**
This PR adds multiple entry points support to traefik, aka, traefik can now listens on http and https :)
This PR also add entry point redirection.

Simple entrypoint definition:

```toml
[entryPoints]
  [entryPoints.http]
  address = ":80"
```

To redirect an http entrypoint to an https entrypoint (with SNI support):
```toml
[entryPoints]
  [entryPoints.http]
  address = ":80"
    [entryPoints.http.redirect]
      entryPoint = "https"
  [entryPoints.https]
  address = ":443"
    [entryPoints.https.tls]
      [[entryPoints.https.tls.certificates]]
      CertFile = "integration/fixtures/https/snitest.com.cert"
      KeyFile = "integration/fixtures/https/snitest.com.key"
      [[entryPoints.https.tls.certificates]]
      CertFile = "integration/fixtures/https/snitest.org.cert"
      KeyFile = "integration/fixtures/https/snitest.org.key"
```

To redirect an entrypoint rewriting the URL:
```toml
[entryPoints]
  [entryPoints.http]
  address = ":80"
    [entryPoints.http.redirect]
      regex = "^http://localhost/(.*)"
      replacement = "http://mydomain/$1"
```

Entrypoints to be used by frontends that do not specify any entrypoint:

```toml
defaultEntryPoints = ["http", "https"]
```

Frontends can overrides `defaultEntryPoints`:

```toml
[frontends]
  [frontends.frontend1]
  backend = "backend2"
    [frontends.frontend1.routes.test_1]
    rule = "Host"
    value = "test.localhost"
  [frontends.frontend2]
  backend = "backend1"
  passHostHeader = true
  entrypoints = ["https"] # overrides defaultEntryPoints
    [frontends.frontend2.routes.test_1]
    rule = "Host"
    value = "{subdomain:[a-z]+}.localhost"
  [frontends.frontend3]
  entrypoints = ["http", "https"] # overrides defaultEntryPoints
  backend = "backend2"
    rule = "Path"
    value = "/test"
```

- [x] multiple entry points
- [x] redirection
- [x] providers
- [x] documentation

Fixes #113 and some of #61

### CODE REVIEW

**Comment by dontrebootme** [comment]
Wow, multiple entrypoint and redirection configuration? This looks great!

**Comment by emilevauge** [comment]
@rebaseTraefik

**Comment by rebaseTraefik** [comment]
I just pushed up the changes, enjoy!

**Comment by vdemeester** [comment]
@rebaseTraefik

**Comment by rebaseTraefik** [comment]
I just pushed up the changes, enjoy!

File: cmd.go [hunk 1]
```diff
@@ -9,6 +9,7 @@ import (
 	"strings"
 	"time"
 
+	"encoding/json"
```
Comment (vdemeester) [inline cmd.go]: One line up :angel:

File: configuration.go [hunk 2]
```diff
@@ -11,17 +11,18 @@ import (
 	"github.com/emilevauge/traefik/types"
 	"github.com/mitchellh/mapstructure"
 	"github.com/spf13/viper"
+	"regexp"
```
Comment (vdemeester) [inline configuration.go]: And up too :angel:

File: configuration.go [hunk 3]
```diff
@@ -33,6 +34,110 @@ type GlobalConfiguration struct {
...
+// The String method's output will be used in diagnostics.
+func (ep *EntryPoints) String() string {
+	return ""
+}
+
+// Set is the method to set the flag value, part of the flag.Value interface.
+// Set's argument is a string to be parsed to set the flag.
+// It's a comma-separated list, so we split it.
+func (ep *EntryPoints) Set(value string) error {
+	regex := regexp.MustCompile("(?:Name:(?P<Name>\\S*))\\s*(?:Address:(?P<Address>\\S*))?\\s*(?:TLS:(?P<TLS>\\S*))?\\s*(?:Redirect.EntryPoint:(?P<RedirectEntryPoint>\\S*))?\\s*(?:Redirect.Regex:(?P<RedirectRegex>\\S*))?\\s*(?:Redirect.Replacement:(?P<RedirectReplacement>\\S*))?")
```
Comment (vdemeester) [inline configuration.go]: Oh boy this is sexy :smiley_cat:

File: integration/marathon_test.go [hunk 4]
```diff
@@ -5,6 +5,7 @@ import (
 	"os/exec"
 	"time"
 
+	"fmt"
```
Comment (vdemeester) [inline integration/marathon_test.go]: One line up :angel:

**Comment by emilevauge** [comment]
Fixed and rebased.
If you don't have any better idea regarding the ugly regex, let's merge it then. Will find a better way to manage args later.

**Comment by vdemeester** [comment]
@rebaseTraefik

**Comment by rebaseTraefik** [comment]
I could not rebase add-multiple-entrypoints-support with master. There are conflicts.

**Comment by rogierlommers** [comment]
Sorry to bump, but I'm having a hard time to create a frontend (only a frontend) which redirects traffic from `abc.com` to `xyz.com`. Can someone please point me to some examples?

**Comment by traefiker** [comment]
Hi! I'm Træfiker :robot: the bot in charge of communication regulation.

Thanks for your interest in Træfik, but this issue is now closed :lock:

You can join our [Slack workspace](https://traefik.herokuapp.com/) for more [community #support](https://traefik.slack.com/messages/C0CDT22PJ/).

If needed, you can also open a new issue (and follow the [bug](https://raw.githubusercontent.com/containous/traefik/master/.github/ISSUE_TEMPLATE/bugs.md) or [feature](https://raw.githubusercontent.com/containous/traefik/master/.github/ISSUE_TEMPLATE/features.md) templates).

**Pull request merged**